controllers/replicaset_controller_test.go

```diff
@@ -296,6 +300,177 @@ func TestService_usesCustomMongodPortWhenSpecified(t *testing.T) {
 	assertReconciliationSuccessful(t, res, err)
 }
 
+func createOrUpdatePodsWithVersions(t *testing.T, c k8sClient.Client, name types.NamespacedName, versions []string) {
+	for i, version := range versions {
+		createPodWithAgentAnnotation(t, c, types.NamespacedName{
+			Namespace: name.Namespace,
+			Name:      fmt.Sprintf("%s-%d", name.Name, i),
+		}, version)
+	}
+}
+
+func createPodWithAgentAnnotation(t *testing.T, c k8sClient.Client, name types.NamespacedName, versionStr string) {
+	pod := corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+			Annotations: map[string]string{
+				"agent.mongodb.com/version": versionStr,
+			},
+		},
+	}
+
+	err := c.Create(context.TODO(), &pod)
+
+	if err != nil && apiErrors.IsAlreadyExists(err) {
+		err = c.Update(context.TODO(), &pod)
+		assert.NoError(t, err)
+	}
+
+	assert.NoError(t, err)
+}
+
+func TestService_changesMongodPortOnRunningCluster(t *testing.T) {
+	mdb := newTestReplicaSet()
+	namespacedName := mdb.NamespacedName()
+
+	const oldPort = automationconfig.DefaultDBPort
+	const newPort = 8000
+
+	mgr := client.NewManager(&mdb)
+
+	r := NewReconciler(mgr)
+
+	t.Run("Prepare cluster and change port", func(t *testing.T) {
```

Comment:
Just an idea (probably for the future) - if we used an [event recorder](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/recorder) in the controller and record when do we change ports, the test would be much simpler. We'd just check if proper events were emitted in proper order.

Reply:
That's nice idea, but wouldn't this make this test to be too coupled with implementation? We don't need to know if we change the service or the pod first but the end result.
Also here we use fake api clients, so we don't have any api machinery in unit tests.

Reply:
There are test event recorder implementation available as well. The idea is to treat it as a "flight recorder". You execute a test, wait till the end result in some sort of `eventually` call and once the result is fine, you analyse the events if things happened in a proper order. This is quite a popular approach in core OpenShift operators.